core/src/providers/openai.rs

```diff
@@ -1015,79 +1222,554 @@ pub async fn chat_completion(
+                                        break 'stream;
+                                    }
+                                }
+                            }
+                        };
 
-    let body = match timeout(Duration::new(60, 0), res.bytes()).await {
-        Ok(Ok(body)) => body,
-        Ok(Err(e)) => Err(e)?,
-        Err(_) => Err(anyhow!("Timeout reading response from OpenAI after 60s"))?,
-    };
+                        // println!("CHUNK: {:?}", chunk);
 
-    let mut b: Vec<u8> = vec![];
-    body.reader().read_to_end(&mut b)?;
-    let c: &[u8] = &b;
+                        // Only stream if choices is length 1 but should always be the case.
+                        match event_sender.as_ref() {
+                            Some(sender) => {
+                                if chunk.choices.len() == 1 {
+                                    // we ignore the role for generating events
 
-    let embeddings: Embeddings = match serde_json::from_slice(c) {
-        Ok(c) => Ok(c),
-        Err(_) => {
-            let error: Error = serde_json::from_slice(c)?;
-            match error.retryable() {
-                true => Err(ModelError {
-                    message: error.message(),
-                    retryable: Some(ModelErrorRetryOptions {
-                        sleep: Duration::from_millis(500),
-                        factor: 2,
+                                    // If we get `content` in the delta object we stream "tokens".
+                                    match chunk.choices[0].delta.get("content") {
+                                        None => (),
+                                        Some(content) => match content.as_str() {
+                                            None => (),
+                                            Some(s) => {
+                                                if s.len() > 0 {
+                                                    let _ = sender.send(json!({
+                                                        "type": "tokens",
+                                                        "content": {
+                                                            "text": s,
+                                                        },
+                                                    }));
+                                                }
+                                            }
+                                        },
+                                    };
+
+                                    // If we have exactly one `tool_call.function.name` in the delta object, stream a "function_call" event.
+                                    if let Some(tool_calls) = chunk.choices[0]
+                                        .delta
+                                        .get("tool_calls")
+                                        .and_then(|v| v.as_array())
+                                    {
+                                        if tool_calls.len() == 1 {
```

Comment:
if we have several, we don't want to stream 1 event per tool call ?

Reply:
Based on my experience, we can have maximum one tool call. The goal here is to implement appropriate logging to capture any instances where our code faces unsupported features/behavior.

Reply:
Ok !